fix(rpc module): close subscription task when a subscription is `unsubscribed` via the `unsubscribe call`

niklasad1 opened the pull request:
The reason for this is that the stream passed to `pipe_from_stream` might not produce new items and that can lead to that calls that are `unsubscribed` is not terminated until an item for the underlying stream is produced.

In worst these tasks will not be terminated until the actual connection is closed which this fixes.

After this PR the subscriptions is terminated once the unsubscribe call is received.

niklasad1 committed:
print subscription info; once per minute

niklasad1 committed:
refactor: remove SubscriptionSink::inner_send

niklasad1 committed:
fix: close running task if unsubscribed

niklasad1 commented on core/src/server/rpc_module.rs:
```diff
@@ -722,6 +722,8 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 					};
 					let sub_id = sub_id.into_owned();
 
+					tracing::info!("unsubscribe=`{}` id={:?}", unsubscribe_method_name, sub_id);
```
```suggestion
```

niklasad1 committed:
Update core/src/server/rpc_module.rs

niklasad1 committed:
Update core/src/server/rpc_module.rs

niklasad1 commented on core/src/server/rpc_module.rs:
```diff
@@ -948,21 +968,22 @@ impl SubscriptionSink {
 
 	fn is_active_subscription(&self) -> bool {
-		Arc::strong_count(&self.active_sub) > 1
+		self.subscribers.lock().contains_key(&self.uniq_sub)
```
this is a bit annoying with extra mutex accesses but I kept for ease of use.

we need another `AtomicBool` or something because `rx side` of oneshot can't be checked

niklasad1 committed:
fix nits

niklasad1 commented on core/src/server/rpc_module.rs:
```diff
@@ -922,11 +938,14 @@ impl SubscriptionSink {
 				Either::Left((Ok(None), _)) => break SubscriptionClosed::Success,
-				Either::Right(((), _)) => {
+				Either::Right((_, _)) => {
+					tracing::info!("subscription stream terminated; by connection reset most likely");
```
```suggestion
```

niklasad1 committed:
Update core/src/server/rpc_module.rs

niklasad1 committed:
add test for canceling subscriptions

niklasad1 committed:
Merge remote-tracking branch 'origin/na-close-tasks-when-unsubscribed' into na-close-tasks-when-unsubscribed

jsdw commented on core/src/server/rpc_module.rs:
```diff
@@ -881,25 +882,40 @@ impl SubscriptionSink {
 			Some(close_notify) => close_notify,
-			None => return SubscriptionClosed::RemotePeerAborted,
+			None => {
+				on_close(SubscriptionClosed::RemotePeerAborted, self);
```
I don't really understand the benefit of having this `on_close` closure versus returning, say, `(SubscriptionClosed, Self)`?

And that makes me wonder; would this method take `&mut self` instead then, so it wouldn't need to return `Self` at all, just `SubscriptionClosed`?

And then I wonder, would it be nicer to remove the `SubscriptionClosed::Success` variant and actually return `Result<(), SubscriptionClosed>` rather than `SubscriptionClosed` (and maybe then rename `SubscriptionClosed` to something like `SubscriptionCloseError`?

jsdw replied:
I had to double check, but what I mean is that the closure doesn't protect against it being reused? It might be a little safer if it was handed `&mut self` instead of `self` though, but the safest approach if people need access to some of the functionality from `self`, but not all of it, would be to wrap `self` in some type to make sure it can only be used in a certain way?

jsdw replied:
Dumb question; why can't it be reused?

niklasad1 replied:
The refactoring [here](https://github.com/paritytech/jsonrpsee/blob/na-close-tasks-when-unsubscribed/core/src/server/rpc_module.rs#L904) which takes the `unsubscribed oneshot` which is why I wanted the `pipe_from_sink(self)` rather than `pipe_from_sink(&mut self)`

So if I do this:

```rust
  sink.pipe_from(stream1).await;
  // this will compile but not work, unsubscribe oneshot is already consumed via `Option::take`
  sink.pipe_from(stream2).await;
```

But yeah I can do nasty things in the exit closure too you correct,

```rust
  sink.pipe_from(stream1, |_, _| { i can do nasty stuff here as well but async stuff right }).await;
  // this won't compile
  sink.pipe_from(stream2).await;
```

niklasad1 replied:
reverted good catch James.

niklasad1 committed:
revert closure stuff

niklasad1 committed:
Revert "print subscription info; once per minute"

This reverts commit 366176a81d95d81d6cc7d2eb0d8b3c949f47815a.

jsdw commented on core/src/server/rpc_module.rs:
```diff
@@ -891,14 +892,26 @@ impl SubscriptionSink {
+			}
+		};
+
+		let sub_closed = match self.unsubscribe.take() {
```
Ah I see, so this oneshot unsubscribe channel is removed here, which means we can only use `pipe_from_try_stream` once.

Is it possible to take a mutable ref to it instead to use below (which may need pinning)? Maybe then this method would still be reusable?

niklasad1 replied:
it won't work because `send` takes a `&self as well.

I changed it to `tokio::sync::watch` for ease of use.

niklasad1 committed:
use tokio::sync::watch instead of oneshot

The receiver is clonable and it's possible to check whether the sender is still alive

dvdplm commented on tests/tests/helpers.rs:
```diff
@@ -108,6 +112,54 @@ pub async fn websocket_server_with_subscription() -> (SocketAddr, WsServerHandle
 
+	module
+		.register_subscription("subscribe_5_ints", "n", "unsubscribe_5_ints", |_, pending, _| {
+			let mut sink = match pending.accept() {
```
maybe it's better to `unwrap()` here?

dvdplm commented on tests/tests/helpers.rs:
```diff
@@ -108,6 +112,54 @@ pub async fn websocket_server_with_subscription() -> (SocketAddr, WsServerHandle
+			"n",
+			"unsubscribe_with_err_on_stream",
+			move |_, pending, _| {
+				let mut sink = match pending.accept() {
+					Some(sink) => sink,
+					_ => return,
+				};
```
same

dvdplm commented on tests/tests/helpers.rs:
```diff
@@ -133,6 +185,34 @@ pub async fn websocket_server() -> SocketAddr {
 	addr
 }
 
+/// Yields at one item then sleeps for an hour.
```
```suggestion
/// Yields one item then sleeps for an hour.
```

niklasad1 committed:
Update tests/tests/helpers.rs

Co-authored-by: David <dvdplm@gmail.com>

niklasad1 committed:
Update core/src/server/rpc_module.rs

Co-authored-by: David <dvdplm@gmail.com>

niklasad1 committed:
grumbles: use unwrap in tests

jsdw commented on tests/tests/helpers.rs:
```diff
@@ -108,6 +100,48 @@ pub async fn websocket_server_with_subscription() -> (SocketAddr, WsServerHandle
 		.unwrap();
 
+	module
+		.register_subscription("subscribe_5_ints", "n", "unsubscribe_5_ints", |_, pending, _| {
```
Now we can call `pipe_from_stream` more than once, should we also have a test to make sure that we can?

jsdw reviewed (approved):
Looks great to me!

niklasad1 committed:
add test for reuse pipe_from_stream
